Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -1592,6 +1593,16 @@ void RenderLayer::setAncestorChainHasVisibleDescendant()
 
 void RenderLayer::updateAncestorDependentState()
 {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (renderer().document().settings().layerBasedSVGEngineEnabled()) {
+        m_enclosingSVGResourceContainer = lineageOfType<RenderSVGResourceContainer>(renderer()).first();
```

Comment:
While this is convenient, we could halve the necessary tree walks, if we manually walk the ancestor chain once, and check for 'RenderSVGResourceContainer' and 'RenderSVGHiddenContainer' at the same time. I think this would be a good argument to do that instead of lineageOfType<>.

Also now that I think about it:

```
<defs>
  <clipPath>
    <rect>
```

The `<rect>` has both m_enclosingSVGResourceContainer and m_enclosingSVGHiddenContainer set. It would be sufficient to only have m_enclosingSVGResourceContainer set. In practice, with the current code, it won't make a difference, however for safety reasons this is a sane choice -- there is no intention to handle a specific case where both are set.

So I suggest to write the ancestor walk manually, and either stop at the first RenderSVGHiddenContainer or RenderSVGResourceContainer -- this way we avoid the two tree walks, and avoid awkward situations where both are set.

Reply:
Done, the stop condition is a bit tricky.